python/pyarrow/io.pxi

```diff
@@ -262,6 +271,165 @@ cdef class NativeFile:
         Alias for read, needed to match the IOBase interface."""
         return self.read(nbytes=None)
 
+    def readall(self):
+        return self.read()
+
+    def readinto(self, b):
+        """
+        Read into the supplied buffer
+
+        Parameters
+        -----------
+        b: any python object supporting buffer interface
+
+        Returns
+        --------
+        number of bytes written
+        """
+
+        cdef:
+            int64_t buf_size
+            int64_t bytes_read
+            uint8_t* buf
+            cp.Py_buffer py_buffer
+
+        self._assert_readable()
+
+        if cp.PyObject_GetBuffer(b, &py_buffer, cp.PyBUF_CONTIG) == -1:
+            raise ValueError("{} doesnot support buffer interface".format(b))
+
+        buf = <uint8_t*> py_buffer.buf
+
+        buf_size = py_buffer.len
+
+        with nogil:
+            check_status(self.rd_file.get().Read(buf_size, &bytes_read, buf))
+
+        cp.PyBuffer_Release(&py_buffer)
+
+        return bytes_read
+
+    def readline(self, size=None):
+        """Read and return a line of bytes from the file.
+
+        If size is specified, read at most size bytes.
+
+        Line terminator is always b"\\n"."""
+
+        cdef:
+            int64_t c_nbytes
+            int64_t bytes_to_read
+            int64_t bytes_read
+            int64_t linebreak_pos = 0
+            int64_t initial_pos
+            uint8_t* linebreak_ptr
+            PyObject* obj
+            shared_ptr[CBuffer] read_buffer
+            int64_t READ_BUFFER_SIZE = 4096
+
+        self._assert_readable()
+
+        if size is None:
+            c_nbytes = self.size() - self.tell()
+        else:
+            c_nbytes = size
+
+        # ReadAt moves position inconsistently, so save the initial position
+        check_status(self.rd_file.get().Tell(&initial_pos))
+
+        with nogil:
+            while linebreak_pos < c_nbytes:
+                if self.rd_file.get().supports_zero_copy():
+                    # we try to read the whole file, since this operation
+                    # is zero copy
+                    check_status(self.rd_file.get().GetSize(&bytes_to_read))
+                else:
+                    # we read only READ_BUFFER_SIZE next bytes which might
+                    # not be the ideal size for every file type
+                    bytes_to_read = READ_BUFFER_SIZE
+                check_status(
+                    self.rd_file.get().ReadAt(initial_pos + linebreak_pos,
```

Comment:
Maybe use `Read` here? This will prevent additional seeks on the underlying file handle

Reply:
Ist there a reason why `Readable::Read(int64_t, shared_ptr<Buffer>)` ist called `ReadB` in Cython? In other places we use method overloading.

Reply:
Hmm, no idea. Can you try fixing it?

Reply:
There's apparently a cython bug which prevents proper overload resolution with multiple layers of inheritance. I put a comment explaining the name change.